File: app/src/main/java/org/gampiot/robok/ui/fragments/project/create/CreateProjectFragment.java
package org.gampiot.robok.ui.fragments.project.create;

import android.os.Bundle;
import android.view.LayoutInflater;
import android.view.View;
import android.view.ViewGroup;

import androidx.annotation.NonNull;
import androidx.annotation.Nullable;
import androidx.activity.OnBackPressedDispatcher;
import androidx.annotation.IdRes;

import com.google.android.material.dialog.MaterialAlertDialogBuilder;
import com.google.android.material.transition.MaterialSharedAxis;

import org.gampiot.robok.R;
import org.gampiot.robok.databinding.FragmentCreateProjectBinding;
import org.gampiot.robok.ui.fragments.editor.EditorFragment;
import org.gampiot.robok.ui.fragments.project.template.model.ProjectTemplate;
import org.gampiot.robok.ui.fragments.project.create.util.ProjectManager;
import org.gampiot.robok.feature.util.base.RobokFragment;
import org.gampiot.robok.feature.util.Helper;

import java.io.File;

public class CreateProjectFragment extends RobokFragment implements ProjectManager.Listener {

    private FragmentCreateProjectBinding binding;
    
    private ProjectTemplate template;
    private ProjectManager projectManager;
    
    public CreateProjectFragment() {
         this.template = defaultProjectTemplate();
    }
    
    public CreateProjectFragment(@NonNull ProjectTemplate template) {
         this.template = template;
    }

    @Nullable
    @Override
    public View onCreateView(@NonNull LayoutInflater inflater, @Nullable ViewGroup container, @Nullable Bundle savedInstanceState) {
         binding = FragmentCreateProjectBinding.inflate(inflater, container, false);
         return binding.getRoot();
    }

    @Override
    public void onViewCreated(@NonNull View view, @Nullable Bundle savedInstanceState) {
         super.onViewCreated(view, savedInstanceState);
         configureToolbarNavigationBack(binding.toolbar);
         
         loadTemplate();
         
         OnBackPressedDispatcher onBackPressedDispatcher = requireActivity().getOnBackPressedDispatcher();
         
         projectManager = new ProjectManager(requireContext());
         projectManager.setListener(this);
         
         var helper = new Helper();
         binding.buttonBack.setOnClickListener(helper.getBackPressedClickListener(onBackPressedDispatcher));
         binding.buttonNext.setOnClickListener(v -> create());
    }

    @Override
    public void onDestroyView() {
         super.onDestroyView();
         binding = null;
    }
    
    public void loadTemplate() {
         binding.projectName.setText(template.name);
         binding.projectPackageName.setText(template.packageName);
    }
    
    public void create () {
         projectManager.create(
               binding.projectName.getText().toString(),
               binding.projectPackageName.getText().toString(),
               template
         );
    }
    
    public ProjectTemplate defaultProjectTemplate () {
         var template = new ProjectTemplate();
         template.setName(getString(org.gampiot.robok.feature.res.R.string.template_name_empty_game));
         template.setPackageName("com.robok.empty");
         template.setZipFileName("empty_game");
         template.setJavaSupport(true);
         template.setKotlinSupport(false);
         template.setImage(R.drawable.ic_empty_game);
         return template;
    }
    
    @Override
    public void onProjectCreate() {
         var dialog = 
             new MaterialAlertDialogBuilder(requireContext())
                  .setTitle(getString(org.gampiot.robok.feature.res.R.string.warning_project_created_title))
                  .setMessage(getString(org.gampiot.robok.feature.res.R.string.warning_project_created_message))
                  .setPositiveButton(getString(org.gampiot.robok.feature.res.R.string.title_open_project), (d, i) -> {
                        openFragment(new EditorFragment(projectManager.projectDir));
                  })
                  .setNegativeButton(getString(org.gampiot.robok.feature.res.R.string.common_word_ok), (dd, ii) -> {
                        dd.dismiss();
                  })
                  .create();
         dialog.show();         
    }
        
    @Override
    public void onProjectCreateError() {
         //
    }
}

File: app/src/main/java/org/gampiot/robok/ui/fragments/project/create/CreateProjectFragment.kt
package org.gampiot.robok.ui.fragments.project.create

/*
 *  This file is part of Robok © 2024.
 *
 *  Robok is free software: you can redistribute it and/or modify
 *  it under the terms of the GNU General Public License as published by
 *  the Free Software Foundation, either version 3 of the License, or
 *  (at your option) any later version.
 *
 *  Robok is distributed in the hope that it will be useful,
 *  but WITHOUT ANY WARRANTY; without even the implied warranty of
 *  MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
 *  GNU General Public License for more details.
 *
 *  You should have received a copy of the GNU General Public License
 *   along with Robok.  If not, see <https://www.gnu.org/licenses/>.
 */ 

import android.os.Bundle
import android.os.Environment
import android.view.LayoutInflater
import android.view.View
import android.view.ViewGroup
import android.content.Intent

import androidx.activity.OnBackPressedDispatcher
import androidx.annotation.NonNull
import androidx.annotation.Nullable

import com.google.android.material.dialog.MaterialAlertDialogBuilder

import org.gampiot.robok.R
import org.gampiot.robok.databinding.FragmentCreateProjectBinding
import org.gampiot.robok.ui.activities.editor.EditorActivity
import org.gampiot.robok.ui.fragments.project.template.model.ProjectTemplate
import org.gampiot.robok.ui.fragments.project.create.util.ProjectManager
import org.gampiot.robok.feature.util.base.RobokFragment
import org.gampiot.robok.feature.util.Helper

import java.io.File

class CreateProjectFragment(
    private val template: ProjectTemplate
) : RobokFragment(), ProjectManager.CreationListener {

    private var binding: FragmentCreateProjectBinding? = null
    private lateinit var projectManager: ProjectManager

    @Nullable
    override fun onCreateView(@NonNull inflater: LayoutInflater, @Nullable container: ViewGroup?, @Nullable savedInstanceState: Bundle?): View {
        binding = FragmentCreateProjectBinding.inflate(inflater, container, false)
        return binding!!.root
    }

    override fun onViewCreated(@NonNull view: View, @Nullable savedInstanceState: Bundle?) {
        super.onViewCreated(view, savedInstanceState)
        configureToolbarNavigationBack(binding!!.toolbar)

        loadTemplate()

        val onBackPressedDispatcher = requireActivity().onBackPressedDispatcher

        projectManager = ProjectManager(requireContext())
        projectManager.setListener(this)

        val helper = Helper()
        binding!!.buttonBack.setOnClickListener(helper.getBackPressedClickListener(onBackPressedDispatcher))
        binding!!.buttonNext.setOnClickListener { create() }
    }

    override fun onDestroyView() {
        super.onDestroyView()
        binding = null
    }

    private fun loadTemplate() {
        binding!!.projectName.setText(template.name)
        binding!!.projectPackageName.setText(template.packageName)
    }

    private fun create() {
        val projectPath = File(Environment.getExternalStorageDirectory(), "Robok/.projects/${binding!!.projectName.text}")
        projectManager.setProjectPath(projectPath)
        projectManager.create(
            binding!!.projectName.text.toString(),
            binding!!.projectPackageName.text.toString(),
            template
        )
    }

    private fun defaultProjectTemplate(): ProjectTemplate {
        return ProjectTemplate().apply {
            name = getString(org.gampiot.robok.feature.res.R.string.template_name_empty_game)
            packageName = "com.robok.empty"
            zipFileName = "empty_game"
            javaSupport = true
            kotlinSupport = false
            imageResId = R.drawable.ic_empty_game
        }
    }

    override fun onProjectCreate() {
         val dialog = MaterialAlertDialogBuilder(requireContext())
             .setTitle(getString(org.gampiot.robok.feature.res.R.string.warning_project_created_title))
             .setMessage(getString(org.gampiot.robok.feature.res.R.string.warning_project_created_message))
             .setPositiveButton(getString(org.gampiot.robok.feature.res.R.string.title_open_project)) { _, _ ->
                 val bundle = Bundle().apply {
                     putString("projectPath", projectManager.getProjectPath().absolutePath)
                 }
              
                 val intent = Intent(requireContext(), EditorActivity::class.java).apply {
                      putExtras(bundle)
                 }
                 startActivity(intent)
             }
             .setNegativeButton(getString(org.gampiot.robok.feature.res.R.string.common_word_ok)) { dialog, _ ->
                 dialog.dismiss()
             }
             .create()
        dialog.show()
    }

    override fun onProjectCreateError() {
    }
}